### ISSUE TITLE
Improve DevTools Focus Rings

### ISSUE THREAD

**Issue by violasong** [devtools-html/ux#7]
## DevTools focus rings that need UX help:
- Tabbing through changeable attributes that already use blue50 to color the entire row, e.g. in Inspector - needs a lighter-colored border, and bolder, maybe 2px to be visible enough.        
- Tabbing through changeable text that doesn't use a dark background color, e.g. in Rules.
- Any borderless buttons and other minimalist controls that still use the dotted-line style, which I'm trying to get rid of - these should either use the 'main tabs' or 'margin-less' style focus rings.

## DevTools focus rings (UX solved, some implementation needed):
- **Simple margin-less fields** at the edge of browser screen or toolbar, such as the Console filter field, use a simple 1px blue border.
**Main tabs** use hover style as focus style. I’d like to extend this to all of the toolbar buttons in each panel
**Large areas** that can be tabbed to and scrolled, such as DOM, should have a "not-focused, selected row" styling, e.g. black-on-gray instead of white-on-blue. (This will be in effect soon for the a11y panel.)
- We would want to use the regular DSG styling for any regular-sized input fields like the checkboxes in Settings

## Design System
- This needs to be better documented with images and CSS for the design system.

## Examples of Firefox focus rings
<img width="459" alt="image" src="https://user-images.githubusercontent.com/55994/44432980-4ac69200-a559-11e8-8e82-da6a28f2e19d.png">
<img width="405" alt="image" src="https://user-images.githubusercontent.com/55994/44432992-56b25400-a559-11e8-9686-4ce89ac866a2.png">

**Comment by fvsch** [comment]
> Main tabs use hover style as focus style. I’d like to extend this to all of the toolbar buttons in each panel.

This can be useful, but is not enough for an accessible focus style, because the hover style is too subtle. For keyboard focus, it’s best to also add a strongly contrasted outline (which can be implemented as a CSS outline, border or box-shadow).

A possibly useful trick is to have all buttons use a 1px transparent border:

```css
.some-devtool-button {
  border: solid 1px transparent;
  border-radius: 3px;
  padding: 2px 4px;
}

.some-devtool-button:-moz-focusring {
  border-color: blue;
}
```

**Comment by jasonLaster** [comment]
In the debugger and i believe the toolbox, we only show the focus ring on key down. We do not show it on mouse down. I think this is a nice pattern that other tools could benefit from.

**Comment by aburone** [comment]
Hi. Which one would be the good focus ring?

**Comment by fvsch** [comment]
The Photon design system [defines focus styles for buttons](https://design.firefox.com/photon/components/buttons.html#default-4) as:

```css
box-shadow: 0 0 0 1px #0a84ff inset, 0 0 0 1px #0a84ff, 0 0 0 4px rgba(10, 132, 255, 0.3)
```

We could use this style, but Photon elements are often 50% bigger than DevTools elements. So maybe we need to adapt it.

Personally, I would love it if each interactive element could have a 1px border (transparent by default, if the design doesn't use a border). This would enable us to do:

1. Focus styles with a strongly contrasted blue border, plus a lighter 1px or 2px blue box-shadow.
2. High contrast theme where all borders are visible and contrasted.

That would be cool, but I'm not sure a high contrast theme is on the books. So implementing with an inset box-shadow (1px) and an "outset" partial opacity box-shadow (1px or 2px) could be okay.

**Comment by aburone** [comment]
OK. Sooo. Do I start hacking around and put a PoC here? 
I've read the proton style guide. Is there one for the devtools (specially the dark theme) 

Shall I create a bug for every instance of inputs with the wrong focus (that would be madness IMHO)

Also I realize that ie RDM styles are totally isolated from other tools. But in the inspector the changes I tried with other bug affected the settings....
Is there any documentation about this behavior?

**Comment by fvsch** [comment]
> Also I realize that ie RDM styles are totally isolated from other tools. But in the inspector the changes I tried with other bug affected the settings....
> Is there any documentation about this behavior?

The code is the documentation. ;)

Some stylesheets are generic and loaded in most all contexts, including:
- `devtools/client/themes/variables.css`
- `devtools/client/themes/common.css`

Some stylesheets are loaded only by a given tool, they can be in the `devtools/client/themes/` folder or in a tool's own folder (`devtools/client/foo`).

**Comment by aburone** [comment]
Perfect. When I have something to show, shall I upload a patch in a bug report? Or is there another preferred way?

**Comment by fvsch** [comment]
I think you can create a bug, e.g. "Explore new focus ring styles in Netmonitor", and reference this issue.
Then you can upload a patch to Phabricator (with `r=fvsch`). I'll note in the bug that we won't merge it.

**Comment by aburone** [comment]
Florens I created a html file with the basic idea and the different triggers.
I used different focus color for dark and light themes.
Everything is defined in variables for easier tinkering.

Let me know how far away of the mark I've got so I can create a better PoC in the actual devtools.
https://bugzilla.mozilla.org/show_bug.cgi?id=1502098

**Comment by violasong** [comment]
Hi @aburone and @fvsch, I'm so happy to see you working on this bug and I would really like to help more with this. I'm currently needing to focus on some tight deadlines coming up so I can't think about this much right now. I'm happy to go with Florens' plan and jump back in with feedback sometime next week when I'm able to. Thanks so much!

**Comment by fvsch** [comment]
We have a design proposal ready for review at https://bugzilla.mozilla.org/show_bug.cgi?id=1502098#c16

**Comment by fvsch** [comment]
Discussed at All Hands with @violasong and :yzen.

Main decisions:

1. The proposal at https://fx-devtools-focusring.netlify.com/ is mostly validate; we'll probably drop the inset shadow for edge-to-edge elements.
2. We do need an implementation that avoids adding focus rings on buttons after mouse clicks. Pending an implementation of `:focus-visible` in Gecko, we may need to experiment with using a polyfill. The Accessibility tool already uses Photon focus styles, and some JS hacks to prevent focus styles after clicks.

One way to go forward is to prototype implementing the new styles and the `:focus-visible` polyfill in the Accessibility tool. I'll change the Bugzilla bug with this goal.
